text/0000-dyn-type.md

```diff
@@ -0,0 +1,2261 @@
+struct Matrix<P: PointerFamily, T> {
+    elements: P::Pointer<[T]>,
+    width: usize,
+    height: usize,
+    stride: usize,
+}
+```
+
+An advantage of GATs is ability to represent multiple pointers, e.g.
+
+```rust ,ignore
+struct Tensor<P: PointerFamily, T> {
+    elements: P::Pointer<[T]>,
+    ranges: P::Pointer<[(/*len*/usize, /*stride*/usize)]>,
+}
+```
+
+and then we get `Tensor<RefFamily<'a>, T>`, `Tensor<BoxFamily, T>`, `Tensor<GcFamily<'root>, T>` etc,
+all of which cannot be represented using custom DST (`&Tensor<T>`, `Box<Tensor<T>>`, `Gc<'root, Tensor<T>>`)
+since the metadata type needs `Copy`.
+
+(Note that such trick does not really depend on GATs.
+For instance, `ndarray` is already using this approach in the [`ArrayBase` type][ndarray::ArrayBase],
+though the implementation is not pretty without GATs.)
+
+[ndarray::ArrayBase]: https://docs.rs/ndarray/0.11.0/ndarray/struct.ArrayBase.html
+
+### Generalized metadata
+
+The bounds for metadata type is currently too strict and too relaxed.
+It is too strict since pointers like `*mut T` does not need `Send + Sync`,
+and references like `&'a T` only needs `'a` lifetime instead of `'static`.
+It is too relaxed since bounds like `UnwindSafe` is missing due to libstd/libcore separation.
+
+Furthermore, as shown above, custom DST is useless for `ndarray`
+as we cannot store the arbitrarily-long stride as a `Vec` is not `Copy`.
+
+These can be solved if we make `Metadata` a GAT which takes a pointer family,
+and allow the pointer family to specify what traits `Metadata` should bound for using associated trait bounds
+([RFC issue #2190]).
+
+```rust ,ignore
+trait PointerFamily {
+    trait MetaBounds;
+    type Pointer<T: ?Sized>;
+}
+
+struct ConstPtrFamily;
+impl PointerFamily for ConstPtrFamily {
+    trait MetaBounds = Copy + Ord + Hash + UnwindSafe + 'static;
+    type Pointer<T: ?Sized> = *const T;
+}
+
+struct RefFamily<'a>(PhantomData<&'a ()>);
+impl<'a> PointerFamily for RefFamily<'a> {
+    trait MetaBounds = Copy + Send + Sync + UnwindSafe + Unpin + 'a;
+    type Pointer<T: ?Sized> = &'a T;
+}
+
+struct RcFamily;
+impl PointerFamily for RcFamily {
+    trait MetaBounds = Clone + UnwindSafe + Unpin + 'static;
+    type Pointer<T: ?Sized> = Rc<T>;
+}
+```
+
+```rust ,ignore
+trait Pointee {
+    type Metadata<P: PointerFamily>: Sized + P::MetaBounds;
+}
+```
+
+and a `Box<Dst>` will be represented as `(Box<()>, Dst::Metadata<BoxFamily>)` in memory.
+This allows us to write the high-dimensional tensor slice as
+
+```rust ,ignore
+impl Pointee for Tensor<T> {
+    type Metadata<P: PointerFamily> = P::Pointer<[(usize, usize)]>;
+}
+```
+
+Since we can’t implement `Pointee` directly, we need to introduce a third kind of DST syntax e.g.
+
+```rust ,ignore
+dyn type Tensor<T>(third_kind<P>: P<[(usize, usize)]>) = [T];
+```
+
+The problem of generalizing `Metadata` is that it will plague every API it touches, e.g.
+
+```rust
+fn metadata<T: ?Sized, P: PointerFamily>(dst: P
```

Comment:
```rust
trait T<U: Align<2>> {
    type V: PartialEq<U> + Align<2>,
}
```

;)

Reply:
🤔 having `Align<2>` could imply a type can simultaneously implement `Align<1> + Align<2> + Align<4> + ...`

Reply:
Oooooops!

```rust
trait T<U: Align<Val = 2>> {
    type V: PartialEq<U> + Align<Val = 2>,
}
```

We already have associated consts so we can probably do it today.